tests/test_metadata_serialization.py

```diff
@@ -25,13 +28,76 @@
     Targets,
     Timestamp,
 )
+from tuf.api.serialization import DeserializationError
 
 logger = logging.getLogger(__name__)
 
 
+# pylint: disable=too-many-public-methods
 class TestSerialization(unittest.TestCase):
     """Test serialization for all classes in 'tuf/api/metadata.py'."""
 
+    invalid_metadata: utils.DataSet = {
+        "no signatures field": b'{"signed": \
+            { "_type": "timestamp", "spec_version": "1.0.0", "version": 1, "expires": "2030-01-01T00:00:00Z", \
+            "meta": {"snapshot.json": {"hashes": {"sha256" : "abc"}, "version": 1}}} \
+        }',
+        "non unique duplicating signatures": b'{"signed": \
+                { "_type": "timestamp", "spec_version": "1.0.0", "version": 1, "expires": "2030-01-01T00:00:00Z", \
+                "meta": {"snapshot.json": {"hashes": {"sha256" : "abc"}, "version": 1}}}, \
+            "signatures": [{"keyid": "id", "sig": "b"}, {"keyid": "id", "sig": "b"}] \
+        }',
+    }
+
+    @utils.run_sub_tests_with_dataset(invalid_metadata)
+    def test_invalid_metadata_serialization(self, test_data: bytes) -> None:
+        # We expect a DeserializationError reraised from ValueError or KeyError.
+        with self.assertRaises(DeserializationError):
+            Metadata.from_bytes(test_data)
+
+    valid_metadata: utils.DataSet = {
+        "multiple signatures": b'{ \
+            "signed": \
+                { "_type": "timestamp", "spec_version": "1.0.0", "version": 1, "expires": "2030-01-01T00:00:00Z", \
+                "meta": {"snapshot.json": {"hashes": {"sha256" : "abc"}, "version": 1}}}, \
+            "signatures": [{ "keyid": "id", "sig": "b"}, {"keyid": "id2", "sig": "d" }] \
+        }',
+        "no signatures": b'{ \
+            "signed": \
+                { "_type": "timestamp", "spec_version": "1.0.0", "version": 1, "expires": "2030-01-01T00:00:00Z", \
+                "meta": {"snapshot.json": {"hashes": {"sha256" : "abc"}, "version": 1}}}, \
+            "signatures": [] \
+        }',
+    }
+
+    @utils.run_sub_tests_with_dataset(valid_metadata)
+    def test_valid_metadata_serialization(self, test_case_data: bytes) -> None:
+        md = Metadata.from_bytes(test_case_data)
+        input_dict = json.loads(test_case_data)
+        self.assertDictEqual(input_dict, md.to_dict())
```

Comment:
This is better as it runs the whole deserialize-pipeline ... It still doesn't test the whole serialize-pipeline or compare the final result to original input.

I can guess why, you are balancing between keeping the input test data readable and being able to compare the input data to expected output of `Metadata.to_bytes()`.

Did you try if the byte array could be kept made equal to expected output (compact mode or not) and still somewhat readable?

Reply:
as mentioned in PR comment, I think we can leave this as is but just plan to add better full-cycle serialization tests in near future